File: FactoryMethod/FM.kt
abstract class Animal {
    abstract fun speak(): String
}

class Dog : Animal() {
    override fun speak(): String {
        return "Woof!"
    }
}

class Cat : Animal() {
    override fun speak(): String {
        return "Meow!"
    }
} 

// Animal?은 Animal 타입이거나 null일 수 있음을 나타낸다.
class AnimalFactory {
    fun createAnimal(animalType: String): Animal? {
        return when (animalType.lowercase()) {
            "dog" -> Dog()
            "cat" -> Cat()
            else -> null
        }
    }
}

// ?. 연산자는 null 안전 호출 연산자로, dog가 null이 아닐 때만 메소드를 호출한다.
fun main() {
    val animalFactory = AnimalFactory()

    val dog = animalFactory.createAnimal("dog")
    val cat = animalFactory.createAnimal("cat")

    println("Dog says: ${dog?.speak()}")
    println("Cat says: ${cat?.speak()}")
}

File: FactoryMethod/FM.cpp
#include <iostream>
#include <string>

using namespace std;

class Animal {
public:
    virtual ~Animal() = default;
    virtual string speak() const = 0;    
};

class Dog : public Animal {
public:
    string speak() const override {
        return "Woof!";
    }
};

class Cat : public Animal {
public:
    string speak() const override {
        return "Meow!";
    }
};

class AnimalFactory {
public:
    static Animal* createAnimal(const string& animalType) {
        if (animalType == "Dog") {
            return new Dog();
        } else if (animalType == "Cat") {
            return new Cat();
        }
        return nullptr; // c++에서는 모든 코드 경로에서 반환값을 명시해야 한다. 즉, dog와 cat이 아닌 경우에 반환값이 정의되어야 한다.
    }
};

int main() {
    Animal* myDog = AnimalFactory::createAnimal("Dog");
    Animal* myCat = AnimalFactory::createAnimal("Cat");

    cout << "Dog says: " << myDog->speak() << endl;
    cout << "Cat says: " << myCat->speak() << endl;

    delete myDog;
    delete myCat;

    return 0;
}